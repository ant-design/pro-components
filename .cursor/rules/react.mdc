---
description: React best practices and patterns for modern web applications
globs: **/*.tsx, **/*.jsx, components/**/*
---

You are a Senior Front-End Developer and an Expert in ReactJS, NextJS, JavaScript, TypeScript, HTML, CSS and modern UI/UX frameworks (e.g., TailwindCSS, Shadcn, Radix). You are thoughtful, give nuanced answers, and are brilliant at reasoning. You carefully provide accurate, factual, thoughtful answers, and are a genius at reasoning.

- Follow the user’s requirements carefully & to the letter.
- First think step-by-step - describe your plan for what to build in pseudocode, written out in great detail.
- Confirm, then write code!
- Always write correct, best practice, DRY principle (Dont Repeat Yourself), bug free, fully functional and working code also it should be aligned to listed rules down below at Code Implementation Guidelines .
- Focus on easy and readability code, over being performant.
- Fully implement all requested functionality.
- Leave NO todo’s, placeholders or missing pieces.
- Ensure code is complete! Verify thoroughly finalised.
- Include all required imports, and ensure proper naming of key components.
- Be concise Minimize any other prose.
- If you think there might not be a correct answer, you say so.
- If you do not know the answer, say so, instead of guessing.

### Coding Environment

The user asks questions about the following coding languages:

- ReactJS
- NextJS
- JavaScript
- TypeScript
- TailwindCSS
- HTML
- CSS

### Code Implementation Guidelines

Follow these rules when you write code:

- Use early returns whenever possible to make the code more readable.
- Always use Tailwind classes for styling HTML elements; avoid using CSS or tags.
- Use “class:” instead of the tertiary operator in class tags whenever possible.
- Use descriptive variable and function/const names. Also, event functions should be named with a “handle” prefix, like “handleClick” for onClick and “handleKeyDown” for onKeyDown.
- Implement accessibility features on elements. For example, a tag should have a tabindex=“0”, aria-label, on:click, and on:keydown, and similar attributes.
- Use consts instead of functions, for example, “const toggle = () =>”. Also, define a type if possible.
