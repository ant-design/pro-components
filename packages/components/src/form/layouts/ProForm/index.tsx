import type { FormProps } from 'antd';
import { Form } from 'antd';
import React from 'react';
import type { CommonFormProps } from '../../BaseForm';
import { BaseForm } from '../../BaseForm';
import { EditOrReadOnlyContext } from '../../BaseForm/EditOrReadOnlyContext';
import { Group, ProFormItem } from '../../components';

export type ProFormProps<
  T = Record<string, any>,
  U = Record<string, any>,
> = Omit<FormProps<T>, 'onFinish'> & CommonFormProps<T, U>;

function ProForm<T = Record<string, any>>(
  props: ProFormProps<T> & {
    children?: React.ReactNode | React.ReactNode[];
  },
) {
  return (
    <BaseForm
      layout="vertical"
      contentRender={(items, submitter) => {
        return (
          <>
            {items}
            {submitter}
          </>
        );
      }}
      {...props}
    />
  );
}

ProForm.Group = Group;
ProForm.useForm = Form.useForm;
ProForm.Item = ProFormItem;
ProForm.useWatch = Form.useWatch;
ProForm.ErrorList = Form.ErrorList;
ProForm.Provider = Form.Provider;
ProForm.useFormInstance = Form.useFormInstance;
ProForm.EditOrReadOnlyContext = EditOrReadOnlyContext;

export { ProForm };
