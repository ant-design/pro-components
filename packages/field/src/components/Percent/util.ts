/** 获取展示符号 */
export function getSymbolByRealValue(realValue: number) {
  if (realValue === 0) {
    return null;
  }
  if (realValue > 0) {
    return '+';
  }
  return '-';
}

/** 获取颜色 */
export function getColorByRealValue(realValue: number /** ,color: string */) {
  if (realValue === 0) {
    return '#595959';
  }
  return realValue > 0 ? '#ff4d4f' : '#52c41a';
}

/** 获取到最后展示的数字 */
export function getRealTextWithPrecision(
  realValue: number,
  precision: number = 2,
) {
  return precision >= 0 ? realValue?.toFixed(precision) : realValue;
}

/**
 * 转化为数字
 * @copy from https://github.com/toss/es-toolkit/blob/32a183828c244d675f46810935e45dfefec81a54/src/compat/util/toNumber.ts#L19
 */
export function toNumber(value: any): number {
  if (typeof value === 'symbol' || value instanceof Symbol) {
    return NaN;
  }

  return Number(value);
}
