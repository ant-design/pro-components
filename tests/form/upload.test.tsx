import ProForm, {
  ProFormUploadButton,
  ProFormUploadDragger,
} from '@ant-design/pro-form';
import { cleanup, fireEvent, render } from '@testing-library/react';
import { Form } from 'antd';
import type { UploadFile } from 'antd/lib/upload/interface';
import { act } from 'react';
import mock from 'xhr-mock';
import { waitForWaitTime } from '../util';

const mockFile = new File(['foo'], 'foo.png', {
  type: 'image/png',
}) as unknown as UploadFile;
const mockFile1 = new File(['foo1'], 'foo1.png', {
  type: 'image/png',
}) as unknown as UploadFile;
const mockFile2 = new File(['foo2'], 'foo2.png', {
  type: 'image/png',
}) as unknown as UploadFile;
export function setup() {
  mock.setup();
  // @ts-ignore
  mock.post('http://upload.com/', (req, res) => {
    req.headers({
      'content-length': '100',
    });
    req.body('thisisbody');
    return res;
  });
}

export const teardown = mock.teardown.bind(mock);

afterEach(() => {
  cleanup();
});

describe('ProFormUpload', () => {
  const errorSpy = vi.spyOn(console, 'error').mockImplementation(() => {});

  beforeEach(() => setup());
  afterEach(() => {
    teardown();
    errorSpy.mockReset();
  });

  it('🏐 ProFormUploadButton support onChange', async () => {
    const fn = vi.fn();
    const onChangeFn = vi.fn();
    const wrapper = render(
      <ProForm
        onValuesChange={(_, values) => {
          fn(values.files);
        }}
      >
        <ProFormUploadButton
          action="http://upload.com"
          listType="text"
          onChange={() => onChangeFn()}
          label="upload"
          name="files"
        />
      </ProForm>,
    );

    act(() => {
      fireEvent.change(
        wrapper.baseElement.querySelector<HTMLDivElement>('.ant-upload input')!,
        {
          target: {
            files: [mockFile],
          },
        },
      );
    });
    await waitForWaitTime(1000);
    expect(fn).toBeCalled();
    expect(onChangeFn).toBeCalledTimes(3);
  });

  it('🏐 ProFormUploadButton support beforeUpload', async () => {
    const fn = vi.fn();
    const wrapper = render(
      <ProForm
        onValuesChange={(_, values) => {
          fn(values.files);
        }}
      >
        <ProFormUploadButton
          action="http://upload.com"
          listType="text"
          label="upload"
          name="files"
          fieldProps={{
            beforeUpload: () => {
              return false;
            },
          }}
        />
      </ProForm>,
    );

    act(() => {
      fireEvent.change(
        wrapper.baseElement.querySelector<HTMLDivElement>('.ant-upload input')!,
        {
          target: {
            files: [mockFile],
          },
        },
      );
    });
    await waitForWaitTime(200);

    act(() => {
      expect(
        wrapper.baseElement.querySelectorAll<HTMLDivElement>(
          'div.ant-upload-list-picture-container',
        ).length,
      ).toBe(0);
    });
  });

  it('🏐 ProFormUploadButton support disable', async () => {
    const wrapper = render(
      <Form>
        <ProFormUploadButton
          disabled
          action="http://upload.com"
          listType="text"
          label="upload"
          name="files"
        />
      </Form>,
    );
    expect(
      wrapper.baseElement.querySelector<HTMLDivElement>('.ant-upload')
        ?.innerHTML,
    ).toMatchSnapshot();
    act(() => {
      wrapper.rerender(
        <Form>
          <ProFormUploadButton
            disabled
            action="http://upload.com"
            listType="text"
            label="upload"
            name="files"
            buttonProps={{
              disabled: true,
              type: 'dashed',
            }}
          />
        </Form>,
      );
    });
    await waitForWaitTime(100);
    expect(
      wrapper.baseElement
        .querySelector<HTMLDivElement>('.ant-upload')
        ?.querySelector('.ant-btn-dashed'),
    ).toBeTruthy();
    act(() => {
      wrapper.rerender(
        <Form>
          <ProFormUploadButton
            disabled={false}
            action="http://upload.com"
            listType="text"
            label="upload"
            name="files"
            buttonProps={{}}
            fieldProps={{
              disabled: true,
            }}
          />
        </Form>,
      );
    });
    await waitForWaitTime(100);
    expect(
      wrapper.baseElement
        .querySelector<HTMLDivElement>('.ant-upload')
        ?.querySelector('.ant-btn-dashed'),
    ).toBeFalsy();
  });

  it('🏐 ProFormUploadDragger support onChange', async () => {
    const fn = vi.fn();
    const onChangeFn = vi.fn();
    const wrapper = render(
      <ProForm
        onValuesChange={(_, values) => {
          fn(values.files);
        }}
      >
        <ProFormUploadDragger
          onChange={() => onChangeFn()}
          action="http://upload.com"
          label="upload"
          name="files"
        />
      </ProForm>,
    );

    act(() => {
      fireEvent.change(
        wrapper.baseElement.querySelector<HTMLDivElement>('.ant-upload input')!,
        {
          target: {
            files: [mockFile],
          },
        },
      );
    });
    await waitForWaitTime(200);
    expect(fn).toBeCalled();
    expect(onChangeFn).toBeCalled();
  });

  it('🏐 ProFormUploadDragger hide when max', async () => {
    const wrapper = render(
      <Form>
        <ProFormUploadDragger
          max={2}
          value={[mockFile, mockFile1, mockFile2]}
          action="http://upload.com"
          label="upload"
          name="files"
        />
      </Form>,
    );

    await waitForWaitTime(200);
    expect(
      getComputedStyle(
        wrapper.baseElement.querySelector<HTMLDivElement>(
          '.ant-upload.ant-upload-drag',
        )!,
      )?.display,
    ).toBe('none');
  });

  it('🏐 ProFormUploadDragger support children', async () => {
    const extra = 'extra';
    const wrapper = render(
      <Form>
        <ProFormUploadDragger
          value={[mockFile, mockFile1, mockFile2]}
          action="http://upload.com"
          label="upload"
          name="files"
        >
          {extra}
        </ProFormUploadDragger>
      </Form>,
    );

    await waitForWaitTime(200);
    expect(
      wrapper.baseElement.querySelector<HTMLDivElement>(
        '.ant-upload-drag .ant-upload-extra',
      )?.textContent,
    ).toBe(extra);
  });

  it('🏐 ProFormUploadButton hide when max', async () => {
    const wrapper = render(
      <Form>
        <ProFormUploadButton
          max={2}
          value={[mockFile, mockFile1, mockFile2]}
          action="http://upload.com"
          label="upload"
          name="files"
        />
      </Form>,
    );

    await waitForWaitTime(200);
    expect(
      wrapper.baseElement.querySelector<HTMLDivElement>(
        '.anticon.anticon-upload',
      ),
    ).toBeFalsy();
  });
});
