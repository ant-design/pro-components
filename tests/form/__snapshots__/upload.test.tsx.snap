// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProFormUpload > 🏐 ProFormUploadButton support disable 1`] = `"<span class="ant-upload ant-upload-disabled"><input name="file" disabled="" type="file" style="display: none;" accept=""><button type="button" class="ant-btn ant-btn-default ant-btn-color-default ant-btn-variant-outlined" disabled=""><span role="img" aria-label="upload" class="anticon anticon-upload"><svg viewBox="64 64 896 896" focusable="false" data-icon="upload" width="1em" height="1em" fill="currentColor" aria-hidden="true"><path d="M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"></path></svg></span><span>单击上传</span></button></span>"`;
